src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -578,15 +620,24 @@ public void Add(T item, EntryState state)
                     }
                     else if (_currentState != state)
                     {
-                        _states = ArrayBuilder<EntryState>.GetInstance(_items.Count - 1, _currentState.Value);
```

Comment:
this approach was an anti pattern.  by doing `GetInstance(_items.Count - 1)` we were always allocating an array one too small for what we needed.  The adding a value would force us to double and then copy and then throw that initial array away.  

The new form both avoids that, and avoids any wastage entirely by ensuring we use _requestedCapacity to get exactly the size we need.